lib/data/external/GCP/GcpApis/mpuHelper.js

```diff
@@ -0,0 +1,218 @@
+            Bucket: params.MPU,
+            Key: tmpKey,
+            MultipartUpload: { Parts: mpuPartList },
+        };
+        return _retryCompose.call(this, composeParams, 0, (err, res) => {
+            if (err) {
+                return cb(err);
+            }
+            mergedObject.VersionId = res.VersionId;
+            mergedObject.ETag = res.ETag;
+            return cb(null, mergedObject);
+        });
+    }, (err, res) => {
+        if (err) {
+            return callback(err);
+        }
+        return callback(null, res.length);
+    });
+}
+
+/**
+ * removeParts - remove all objects created to perform a multipart upload
+ * @param {object} params - remove parts params
+ * @param {string} params.Bucket - bucket name
+ * @param {string} params.MPU - mpu bucket name
+ * @param {string} params.Overflow - overflow bucket name
+ * @param {string} params.Key - object key
+ * @param {string} params.UploadId - MPU upload id
+ * @param {function} callback - callback function to call
+ * @return {undefined}
+ */
+function _removeParts(params, callback) {
+    const _getObjectVersions = (bucketType, callback) => {
+        logger.trace(`remove all parts ${bucketType} bucket`);
+        let partList = [];
+        let isTruncated = true;
+        let nextMarker;
+        const bucket = params[bucketType];
+        return async.whilst(() => isTruncated, moveOn => {
+            const listParams = {
+                Bucket: bucket,
+                Prefix: params.Prefix,
+                Marker: nextMarker,
+            };
+            return this.listVersions(listParams, (err, res) => {
+                if (err) {
+                    logHelper(logger, 'error', 'error in ' +
+                        `removeParts - listVersions ${bucketType}`, err);
+                    return moveOn(err);
+                }
+                nextMarker = res.NextMarker;
+                isTruncated = res.IsTruncated;
+                partList = partList.concat(res.Versions);
+                return moveOn();
+            });
+        }, err => callback(err, partList));
+    };
+
+    const _deleteObjects = (bucketType, partsList, callback) => {
+        logger.trace(`successfully listed ${bucketType} parts`, {
+            objectCount: partsList.length,
+        });
+        const delSlices = createDelSlices(partsList);
+        const bucket = params[bucketType];
+        return async.each(delSlices, (list, moveOn) => {
+            const delParams = {
+                Bucket: bucket,
+                Delete: { Objects: list },
+            };
+            return this.deleteObjects(delParams, err => {
+                if (err) {
+                    logHelper(logger, 'error',
+                        `error deleting ${bucketType} object`, err);
+                }
+                return moveOn(err);
+            });
+        }, err => callback(err));
+    };
+
+    return async.parallel([
+        done => async.waterfall([
```

Comment:
Not a huge concern, but a slightly better approach that avoids having to create the whole list of parts to delete first in memory, thus having slightly better latency and memory usage, could be to use a task queue (`async.queue`) to trigger deletes of parts as soon as the partial listing returns them, while ensuring a limit in the number of concurrent deletes sent.

But for flow control it may be a bit tricky to get right (i.e. trigger the listings to the rate at which we can delete, or vice-versa), and I'm fine with how it's done now anyway, just FYI (we may consider it as a later improvement).

Reply:
Will attempt at `async.queue` in the MPU helper class test implementation